docs/specs/workflow-customization.md

```diff
@@ -0,0 +1,1570 @@
+                            }
+                          },
+                          "signup_flow": {
+                            "type": "object",
+                            "required": ["id"],
+                            "properties": {
+                              "id": { "type": "string", "minLength": 1 }
+                            }
+                          },
+                          "login_flow": {
+                            "type": "object",
+                            "required": ["id"],
+                            "properties": {
+                              "id": { "type": "string", "minLength": 1 }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            ]
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+### JSON Schema of `reauth_flows`
+
+```json
+{
+  "properties": {
+    "reauth_flows": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "required": ["id", "steps"],
+        "properties": {
+          "id": { "type": "string", "minLength": 1 },
+          "steps": {
+            "type": "array",
+            "items": {
+              "type": "object",
+              "required": ["type"],
+              "properties": {
+                "id": { "type": "string", "minLength": 1 },
+                "if": { "type": "string", "format": "x_expression"} },
+                "type": { "type": "string", "enum": ["authentication_method"] }
+              },
+              "allOf": [
+                {
+                  "if": {
+                    "properties": {
+                      "type": { "const": "authentication_method" }
+                    }
+                  },
+                  "then": {
+                    "properties": {
+                      "one_of": {
+                        "type": "array",
+                        "items": {
+                          "type": "object",
+                          "required": ["authentication_method"],
+                          "properties": {
+                            "authentication_method": {
+                              "type": "object",
+                              "required": ["id"],
+                              "properties": {
+                                "id": { "type": "string", "minLength": 1 }
+                              }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              ]
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+### Alternative design if Expressions were not used
+
+Here we show how would the configuration looks like if expressions were not used.
+We take the [Use case example 5](#use-case-example-5-comprehensive-example) and rewrite it.
```

Comment:
@chpapa This is the alternative design without IFs. What do you think? This maybe harder to express the conditions of which a authentication method is needed

Reply:
@fungc-io @louischan-oursky I think these look prettier and easier to understand. Did I miss anything? 

My syntax is sth like this:

- `steps` is an array with:
   - `one_of` or `any_of`
   - `type`
- `one_of` or `any_of` is an array with `type`
- `type` is a dictionary with `id` and `steps`

```
signup_flows:
- id: default_signup_flow
  steps:
  - one_of:
    - type: identification
      id: email
      steps:
      - type: authentication
        id: primary_email_code
      - type: verification
        id: email
      - type: authentication
        id: primary_password
      - type: authentication
        id: secondary_totp
    - type: identification
      id: oauth
```

(I noticed that in the config, verification doesn't have id, I put sth there, as I don't feel right it could "imply" which verification it's doing...?)

```
login_flows:
- id: default_login_flow
  steps:
  - one_of:
    - type: identification
      id: email
      steps:
      - one_of:
        - type: authentication
          id: primary_password
          steps:
          - one_of:
            - type: authentication
              id: secondary_totp
            - type: authentication
              id: recovery_code
            - type: authentication
              id: device_token
        - type: authentication
          id: primary_email_code
          steps:
          - one_of:
            - type: authentication
              id: secondary_totp
            - type: authentication
              id: recovery_code
            - type: authentication
              id: device_token
        - type: authentication
          id: primary_passkey
    - type: identification
      id: oauth
    - type: identification
      id: passkey
```

Frankly I like it much better XD I think it would also easier to code a Visual Editor for the flow in future, if our config is like this instead of expression literal?

A few other tweak in my mind:
1. Use `auth`, `identify` and `verify` instead of `authentication`, `identification` and `verification` to save a few keystrokes and easier to read?
2. Use `steps`, `steps_one_of`, `steps_all_of` instead of two layer? Would it be easier to read?

Reply:
According to your proposed syntax, your first example should look like this.

```yaml
# Enumerated flows:
# 1. oauth
# 2. email -> verify the email -> setup OTP authenticator (this is skipped because we have verified the email) -> setup TOTP
# 3. email -> verify the email -> setup password -> setup TOTP

signup_flows:
- id: default_signup_flow
  steps:
    one_of:
    - type: identification
      id: oauth
    - type: identification
      id: email
      steps:
        one_of:
        - type: verification
          id: email
          steps:
            one_of:
            - type: authentication
              id: primary_email_code
              steps:
                one_of:
                - type: authentication
                  id: secondary_totp
            - type: authentication
              id: primary_password
              steps:
                one_of:
                - type: authentication
                  id: secondary_totp
```

One issue of this proposed syntax is the existence of `any_of`.
First, assume we agree on the meaning of `one_of`, `all_of`, and `any_of` according to https://json-schema.org/understanding-json-schema/reference/combining.html and https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/.
Second, assume we have the following configuration.

```yaml
signup_flows:
- id: default_signup_flow
  steps:
    one_of:
    - type: identification
      id: email
      steps:
        one_of:
        - type: verification
          id: email
          steps:
            any_of: # any_of
            - type: authentication
              id: primary_password
              steps:
                one_of:
                - type: authentication
                  id: secondary_totp
            - type: authentication
              id: primary_email_code
              steps:
                one_of:
                - type: authentication
                  id: secondary_totp
```

Since `any_of` meaning one or more, we need to ask the end-user if they want to stop after fulfilling one step.
But this kind of asking is uncommon in signup flows.
Good UX should not make the end-user to make a decision, but this syntax allows this to happen.

Another problem is that how should the nested steps be executed if more than 1 steps in `any_of` were taken?
I do not know the answer because the configuration is ambiguous.

Yet another problem is that without conditionals, we cannot express an optional step.
For example, suppose we want to support a new type of step `prompt_setup_passkey`. (We do have this step in our default UI now).
We cannot express this in the proposed syntax, but we can express this with conditionals.

```
login_flows:
- id: default_login_flow
  steps:
  # ...
  - type: prompt_setup_passkey
    if: !user.has_passkey
```

---

> (I noticed that in the config, verification doesn't have id, I put sth there, as I don't feel right it could "imply" which verification it's doing...?)

`type: verification` uses reference to which step the verification is for.

```yaml
- type: identification_method
  id: setup_identity
  one_of:
   identification_method:
     id: email
- type: verification
  step:
    id: setup_identity
```

---

> Use auth, identify and verify instead of authentication, identification and verification to save a few keystrokes and easier to read?

I agree. Let's change to `authenticate`, `identify`, and `verify`. All of them are verbs and are not in short forms.

Reply:
@louischan-oursky ah I didn't understand the original sample 5; I now understand it better with your comments.

However, I was actually kinda assuming `steps` would perform an action on the array's `type` step by step, so the two `one_of` could be skipped in the `signup_flows`?

```yaml
signup_flows:
- id: default_signup_flow
  steps:
  - one_of:
    - type: identify
      id: oauth
    - type: identify
      id: email
      steps:
      - type: verify
        id: email
      - steps:
        - one_of:
          - type: authenticate
            id: primary_email_code
            steps:
            - type: authenticate
              id: secondary_totp
          - type: authenticate
            id: primary_password
            steps:
            - type: authenticate
              id: secondary_totp
```

Um, actually I thought `any_of` could be for those optional steps; Maybe that's too implicit or complicated. Let's remove it; In that case there are only `steps` and `one_of`, I think we can put them into "same level"? So it should further simplify the redundant steps followed by one_of patterns:

```yaml
signup_flows:
- id: default_signup_flow
  one_of:
  - type: identify
    id: oauth
  - type: identify
    id: email
    steps:
    - type: verify
      id: email
      one_of:
      - type: authenticate
        id: primary_email_code
        steps:
        - type: authenticate
          id: secondary_totp
      - type: authenticate
        id: primary_password
        steps:
        - type: authenticate
          id: secondary_totp
```

Besides `prompt_setup_passkey`, do we have any other optional items in our use case now? If not I think we could:

1. Make the `prompt_setup_passkey` type of action optional by default when the user already have passkey (even make it more explicit say `prompt_setup_passkey_if_not_exists`)
2. We could still have a simple expression in the type configuration or in the flow configuration, but at least we could make sure they're only used in certain edge case; Hence most of the flow would still be eumunerated. That should still make the "flow editor" (if it ever happens) very easy to code?

Reply:
> Make the prompt_setup_passkey type of action optional by default when the user already have passkey (even make it more explicit say prompt_setup_passkey_if_not_exists)

I tend not to couple the condition with the action. Maybe the developer wants to prompt setup passkey when the user did not use passkey to sign in.
Or sometimes the developer wants to prompt setup passkey only when the user never did that before. Having conditionals allow us to support ONE type once and for all.

> We could still have a simple expression in the type configuration or in the flow configuration, but at least we could make sure they're only used in certain edge case; Hence most of the flow would still be eumunerated. That should still make the "flow editor" (if it ever happens) very easy to code?

My concern is I want to follow one principle from the Zen of Python.

> There should be one-- and preferably only one --obvious way to do it.

If we have conditionals only, then it is the only way to do it.
If we have nested steps **AND** conditionals, then we will have 2 ways to do it.

For enumerability, in the current design of only one level of steps and conditionals, we can still enumerate all the flows by enumerating the truth table of all conditionals in the steps.

In case of the "flow editor", the configuration written against the current design does not have duplication. This makes modification and deletion simple to implement too.

Reply:
> I tend not to couple the condition with the action. Maybe the developer wants to prompt setup passkey when the user did not use passkey to sign in.
Or sometimes the developer wants to prompt setup passkey only when the user never did that before. Having conditionals allow us to support ONE type once and for all.

But actually I think we have many of these implicit meanings in various actions already, as our config is basically specifying "what needs to check" on each step, the actual implementation of each action depends on the UI, I think it is unavoidable we will have a lot of reasons why the UX of what the users actually do would depends on the frontend implementation.

I think in your current design, it also assume something like that? (e.g. verify a phone no. after authenticate with SMS, I think we will have a logic to skip / auto satisfy the phone verification?)

> There should be one-- and preferably only one --obvious way to do it.

I don't think that's a valid concern, the expression I proposed here is just for skipping an optional step (And we might not even need that if we have no other optional steps we can think of now besides `prompt_setup_passkey`), not for doing conditions.

Any other cons of using nested configuration? Or pros of using such linear one with complicated expressions?

I think the pros are quite obvious:
1. Easier to implement (just a simple tree in array), both for the config and convert from/to flow chart/flow editor (writing a parser to parse the expression and covert them to tree form is not as simple, I'm not sure why you said "simple to implement")
2. Easier to write the config (ask any developers who love having expressions in declarative config instead of imperative config?)

p.s. if we really need "real conditions" in future, I could imagine we can still implement it via something like this?

```yaml
steps:
  - if: <expressions>
    true:
    false:
```

Reply:
> I think in your current design, it also assume something like that? (e.g. verify a phone no. after authenticate with SMS, I think we will have a logic to skip / auto satisfy the phone verification?)

The point of the configuration is to specify the interaction taken by the end-user.
So each step corresponds to an interaction between the UI and the end-user.
Your example is about whether we mark a phone number as verified every time an OTP is verified for that phone number.
Theoretically this behavior can be attached to a step, thus can be turned on / off for a specific step.

> Any other cons of using nested configuration?

Nested configuration inherently contains duplications.
The editor needs to take care of duplications in case of update / deletion.

> Easier to implement (just a simple tree in array), both for the config and convert from/to flow chart/flow editor (writing a parser to parse the expression and covert them to tree form is not as simple, I'm not sure why you said "simple to implement")

I agree writing a parser to parse the expression is not simple, but it is not very hard neither.
I doubt why we need to convert our linear configuration to a tree.
For enumeration purpose, we can just show all linear configurations with conditionals evaluated.
Or we can draw a flow chart like [this](https://mermaid.live/edit#pako:eNplj0EOgyAQRa9CZl0T3bpoUqm9gF1VXEwEq4mAoZCmMd69qKXadAEh89_MG0aoNReQQtPrZ92isUwRUlgxnMrlrkgUHQmNR6oV72ynFfbxNEM09hG5GicWZKazpSertviC_WPL6ZLTKjgCvzqSvSNZHcmfYz9jjTdHrnjpz3d-YH9COIAURmLH_bfHGWVgWyEFg9Q_uWjQ9ZYBU5NH3cDRityvpQ2kzaw6ADqri5eqIbV-tQCdO7wblKE4oLppLT9N0xv-73S_)

> Easier to write the config (ask any developers who love having expressions in declarative config instead of imperative config?)

Maybe we should create a poll to survey this. In my personal experience, I like the design of GitHub Actions (1 level of steps with conditionals), which I borrow the syntax from.